Add address validation token blog post (#464)
n0-computer/iroh.computer: src/app/blog/address-validation-tokens/page.mdx

import { BlogPostLayout } from '@/components/BlogPostLayout'
import { MotionCanvas } from '@/components/MotionCanvas'

export const post = {
  draft: false,
  author: 'Rüdiger Klaehn',
  date: '2026-06-08',
  title: 'Address validation tokens - what are they good for anyway?',
  description: 'A closer look at QUIC NEW_TOKEN tokens, why they exist, and what they buy you in iroh and noq.',
}

export const metadata = {
    title: post.title,
    description: post.description,
    openGraph: {
      title: post.title,
      description: post.description,
      images: [{
        url: `/api/og?title=Blog&subtitle=${post.title}`,
        width: 1200,
        height: 630,
        alt: post.title,
        type: 'image/png',
      }],
      type: 'article'
    }
}

export default (props) => <BlogPostLayout article={post} {...props} />

> QUIC is not a simple protocol. Most, possibly all of the complexity is there for a good reason. It's a distillation of decades of experience with real world traffic on the internet. But still it can be daunting sometimes.

Reviewer: This is a bit aggressive to me. My attempt would be

> QUIC is not a simple protocol. It is a distillation of decades of experience with real world traffic on the internet. But still, it can be1 daunting sometimes.

Part of the complexity is various mechanisms where one side sends opaque data to the remote that is to be presented in a subsequent packet. Another very complex part is throttling.

During preparation for 1.0 we went over all features of iroh and noq, including some that you rarely interact with as a user. This blog post explains a mechanism that touches both of these topics.

# Anti-amplification

Unlike simpler protocols, QUIC tries very hard to be a good internet citizen and prevent [amplification attacks]. It must not be possible to get a powerful server running a QUIC endpoint to send a lot of data to an unrelated endpoint.

![Reflection-amplification attack: an attacker spoofs the victim's IP, sending tiny requests to reflector services that respond with much larger packets to the victim.](/blog/address-validation-tokens/amplification-attack.svg)

So as long as we have not validated the IP address of the remote, we are only allowed to send a tiny trickle of data to the remote address. To be precise, 3x the amount of data we seemingly received from the remote address.

> For a typical initial [`ClientHello`][clienthello], we receive 1200 bytes (the initial packet is padded to 1200 bytes), so we are allowed to send 3600 bytes.

Reviewer: ```suggestion
For a typical initial [`ClientHello`][clienthello], the server receives 1200 bytes (the initial packet is padded to 1200 bytes), so it is allowed to send 3600 bytes.
```

This, unfortunately, can happen even today. For traditional client/server QUIC connections the rest of the server response to the initial `ClientHello` includes the entire certificate chain. A short chain solely consisting of ECDSA certificates comfortably fits, but e.g. a chain containing multiple RSA-2048 certs does not.

Once you go to post-quantum signatures, even a single ML-DSA-65 won't fit, and a typical chain will be much larger.

There are also extensions that might push the server response over 3600 bytes.

# Trying it out.

To try this out I wrote a small noq demo that just connects to publicly available servers and logs the interaction in [qlog] format. None of this is specific to noq, it is a QUIC mechanism.

![Happy case: server response fits in 3600 bytes](/blog/address-validation-tokens/happy-case.svg)

A real handshake that fits, against `www.cloudflare.com`. Every same-direction packet gap is sub-20 ms; the server's handshake packets arrive in one tight burst.

| time (ms) |    Δ same dir | dir  | space     | pn | size |
|----------:|--------------:|:-----|:----------|---:|-----:|

> |     32.49 |          —    | send | initial   |  0 | 1200 |

Reviewer: Maybe make `dir` `c -> s` and `s -> c`? You're not very clear here on who's perspective you'd describing.

|     51.16 |          —    | recv | initial   |  0 |  155 |

> |     51.16 |       +  0.00 | recv | handshake |  1 | 1200 |

Reviewer: Fun, our qlog accounting is wrong here, because initial and handshake should have been coalesced which means the total datagram would have been 1200 bytes but it shows the handshake as 1200 bytes and another 155 bytes for the initial, which would have been 1155 bytes. And it is the *Initial* packet that needs to be in a 1200 byte datagram, not the handshake one.

|     51.38 |       + 18.89 | send | initial   |  1 |   60 |
|     51.38 |       +  0.00 | send | handshake |  0 | 1140 |
|     54.02 |       +  2.86 | recv | handshake |  2 | 1200 |
|     54.02 |       +  0.00 | recv | handshake |  3 |  835 |
|     54.78 |       +  3.40 | send | handshake |  1 |  115 |
|     54.78 |       +  0.00 | send | 1RTT      |  0 | 1326 |
|     57.17 |       +  2.39 | send | 1RTT      |  1 |  102 |
|     72.20 |       + 18.19 | recv | 1RTT      |  4 |  573 |

[Raw qlog](/blog/address-validation-tokens/handshake-no-stall.qlog)

![Stall case: server response exceeds 3600 bytes](/blog/address-validation-tokens/stall-case.svg)

A real handshake that stalls, against `quic.aiortc.org`. The server sends three Handshake-space packets back to back, hits the 3600-byte cap, and pauses — visible as the **+63.14 ms** gap on the `recv handshake 4` row, exactly one RTT to aioquic.

| time (ms) |    Δ same dir | dir  | space     | pn | size |
|----------:|--------------:|:-----|:----------|---:|-----:|
|     34.67 |          —    | send | initial   |  0 | 1200 |
|    103.71 |          —    | recv | initial   |  0 |  144 |

> |    103.71 |       +  0.00 | recv | handshake |  1 | 1056 |

Reviewer: huh, but here the accounting is correct. something is weird in the other table. also i don't really see what you're trying to show. this might need better visualisation.

Author: This is going against 2 endpoints that have different QUIC impls. Maybe that's the reason. The server side is *not* noq.

Author: The real visualization is the svg diagrams, this is just a qlog from the real world to show this actually happens.

If our server is on the other side of the world, this can make the difference between almost unnoticeable (250ms) and very visibly annoying (500ms).

# Address validation tokens

Now we have established that while in most cases the limit is fine for iroh, we have a concrete scenario where we don't get the benefit of 0-rtt due to the response hitting the 3x anti amplification limit. So what can we do?

It turns out that QUIC contains a mechanism for just this case: [address validation tokens][rfc9000-avt].

On completion of a handshake, we *know* that the remote IP address is valid. We have sent data to it and have received a response *specifically to that data*. So at that point QUIC gives the option to issue tokens that attest for some time that the address is valid and send them to the remote.

When we receive a valid token for a new connection attempt, we know that we have talked to this IP address in the recent past and can lift the 3x / 3600 byte limit.

> We [should][rfc9000-avt] make sure that the tokens aren't reused though. The most trivial implementation would be a set of all non-expired tokens. But this could grow to an arbitrary size, so noq uses a more clever two stage [bloom filter] based approach that has a constant size, at the expense of a very low but non-zero probability of valid tokens being rejected.

Reviewer: ```suggestion
We [SHOULD][rfc9000-avt] make sure that the tokens aren't reused though. The most trivial implementation would be a set of all non-expired tokens. But this could grow to an arbitrary size, so noq uses a more clever two stage [bloom filter] based approach that has a constant size, at the expense of a very low but non-zero probability of valid tokens being rejected.
```
